controller/rancher/certificate_fetcher.go

```diff
@@ -0,0 +1,317 @@
+	filePath := path.Join(relativePath, fileName)
+
+	absFilePath, err := filepath.Abs(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("Error forming path to file %s, error: %v", filePath, err)
+	}
+
+	fInfo, err := os.Lstat(absFilePath)
+	if os.IsNotExist(err) {
+		return nil, fmt.Errorf("File %s does not exist", absFilePath)
+	}
+
+	targetPath := absFilePath
+	if fInfo.Mode()&os.ModeSymlink != 0 {
+		//read symlink
+		targetPath, err := filepath.EvalSymlinks(absFilePath)
+		if err != nil {
+			return nil, fmt.Errorf("File %s pointed by symlink %s does not exist, error: %v", targetPath, absFilePath, err)
+		}
+	}
+	//read target file
+	return fetcher.readFile(targetPath)
+}
+
+func (fetcher *RCertificateFetcher) readFile(filePath string) (*[]byte, error) {
+	contentBytes, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading file %s, error: %v", filePath, err)
+	}
+	return &contentBytes, nil
+}
+
+func (fetcher *RCertificateFetcher) addWatches() {
+	if fetcher.CertDir != "" {
+		err := fetcher.Watcher.Add(fetcher.CertDir)
+		if err != nil {
+			logrus.Errorf("Failed to add watch for Dir %v", fetcher.CertDir)
+		}
+	}
+
+	for dirName := range fetcher.CertsCache {
+		err := fetcher.Watcher.Add(dirName)
+		if err != nil {
+			logrus.Errorf("Failed to add watch for Dir %v", err)
+		}
+	}
+}
+
+func (fetcher *RCertificateFetcher) removeWatches() {
+	if fetcher.CertDir != "" {
+		err := fetcher.Watcher.Remove(fetcher.CertDir)
+		if err != nil {
+			logrus.Errorf("Failed to remove watch for Dir %v", fetcher.CertDir)
+		}
+	}
+
+	for dirName := range fetcher.CertsCache {
+		err := fetcher.Watcher.Remove(dirName)
+		if err != nil {
+			logrus.Errorf("Failed to remove watch from Dir %v", err)
+		}
+	}
+}
+
+func (fetcher *RCertificateFetcher) ProcessFileUpdateEvents(doOnEvent func(string)) {
+	for {
+		logrus.Debugf("Polling for fsnotify events:")
+		select {
+		case event := <-fetcher.Watcher.Events:
+			logrus.Debugf("event: %v", event)
+			if strings.Contains(event.Name, CertName) || strings.Contains(event.Name, KeyName) {
+				if event.Op&fsnotify.Create == fsnotify.Create {
+					logrus.Debugf("Processing event for file: %v", event.Name)
+					lastIndexOfSlash := strings.LastIndex(event.Name, "/")
+					if lastIndexOfSlash != -1 {
+						certDirPath := event.Name[0:lastIndexOfSlash]
+						//reload this cert in cache
+						fetcher.mu.Lock()
+						fetcher.TempCertsMap = make(map[string]*config.Certificate)
+						err := filepath.Walk(certDirPath, fetcher.readCertificate)
+						if err != nil {
+							logrus.Errorf("Error %v reading certDirPath  %v", err, certDirPath)
+						} else {
+							for path, newCert := range fetcher.TempCertsMap {
+								fetcher.CertsCache[path] = newCert
+								logrus.Debugf("Cert is reloaded in cache : %v", newCert)
+							}
+						}
+						fetcher.mu.Unlock()
+					}
+					//scheduleApplyConfig
+					doOnEvent("")
```

Comment:
we should only schedule apply config when reading from the directory was successful.